runtime/integration-tests/src/generic/runtime.rs

```diff
@@ -0,0 +1,151 @@
+	AccountId, Address, AuraId, Balance, BlockNumber, CollectionId, Header, Index, ItemId, LoanId,
+	Moment, PoolId, Signature, TrancheId,
+};
+use cfg_types::{
+	permissions::{PermissionScope, Role},
+	tokens::{CurrencyId, CustomMetadata, TrancheCurrency},
+};
+use codec::Codec;
+use cumulus_primitives_core::PersistedValidationData;
+use cumulus_primitives_parachain_inherent::ParachainInherentData;
+use cumulus_test_relay_sproof_builder::RelayStateSproofBuilder;
+use fp_self_contained::{SelfContainedCall, UncheckedExtrinsic};
+use frame_support::{
+	dispatch::{
+		DispatchClass, DispatchInfo, GetDispatchInfo, Pays, PostDispatchInfo,
+		UnfilteredDispatchable,
+	},
+	inherent::{InherentData, ProvideInherent},
+	traits::{Get, IsType},
+	weights::WeightToFee as _,
+	Parameter,
+};
+use frame_system::{ChainContext, RawOrigin};
+use pallet_transaction_payment::CurrencyAdapter;
+use runtime_common::{
+	apis,
+	fees::{DealWithFees, WeightToFee},
+};
+use sp_io::TestExternalities;
+use sp_runtime::{
+	traits::{AccountIdLookup, Block, Checkable, Dispatchable, Extrinsic, Lookup, Member},
+	ApplyExtrinsicResult, DispatchResult,
+};
+use sp_timestamp::Timestamp;
+
+/// Kind of runtime to check in runtime time
+pub enum RuntimeKind {
+	Development,
+	Altair,
+	Centrifuge,
+}
+
+/// Runtime configuration
+pub trait Runtime:
+	Send
+	+ Sync
+	+ frame_system::Config<
+		Index = Index,
+		AccountId = AccountId,
+		RuntimeCall = Self::RuntimeCallExt,
+		RuntimeEvent = Self::RuntimeEventExt,
+		BlockNumber = BlockNumber,
+		Lookup = AccountIdLookup<AccountId, ()>,
+	> + pallet_pool_system::Config<
+		CurrencyId = CurrencyId,
+		Balance = Balance,
+		PoolId = PoolId,
+		TrancheId = TrancheId,
+	> + pallet_balances::Config<Balance = Balance>
+	+ pallet_investments::Config<InvestmentId = TrancheCurrency, Amount = Balance>
+	+ pallet_pool_registry::Config<
+		CurrencyId = CurrencyId,
+		PoolId = PoolId,
+		Balance = Balance,
+		ModifyPool = pallet_pool_system::Pallet<Self>,
+		ModifyWriteOffPolicy = pallet_loans::Pallet<Self>,
+	> + pallet_permissions::Config<Role = Role, Scope = PermissionScope<PoolId, CurrencyId>>
+	+ pallet_loans::Config<
+		Balance = Balance,
+		PoolId = PoolId,
+		CollectionId = CollectionId,
+		ItemId = ItemId,
+	> + orml_tokens::Config<CurrencyId = CurrencyId, Balance = Balance>
+	+ orml_asset_registry::Config<
+		AssetId = CurrencyId,
+		CustomMetadata = CustomMetadata,
+		Balance = Balance,
+	> + pallet_uniques::Config<CollectionId = CollectionId, ItemId = ItemId>
+	+ pallet_timestamp::Config<Moment = Moment>
+	+ pallet_aura::Config<Moment = Moment, AuthorityId = AuraId>
+	+ pallet_authorship::Config
+	+ pallet_treasury::Config<Currency = pallet_balances::Pallet<Self>>
+	+ pallet_transaction_payment::Config<
+		WeightToFee = WeightToFee,
+		OnChargeTransaction = CurrencyAdapter<pallet_balances::Pallet<Self>, DealWithFees<Self>>,
+	> + pallet_restricted_tokens::Config<
+		Balance = Balance,
+		NativeFungible = pallet_balances::Pallet<Self>,
```

Comment:
IIUC, we should add all pallet's here which we want to use in integration tests. I wonder if the compiler would not throw if runtimes diverge in the Configs they implement or would this be solved by [your efforts to pick the runtime(s) on the test granularity](https://github.com/centrifuge/centrifuge-chain/pull/1583/files#r1348506233)?

Reply:
> IIUC, we should add all pallet's here which we want to use in integration tests

Yes, if you want to use *explicitly* a pallet, then you should add it there.

> I wonder if the compiler would not throw if runtimes diverge in the Configs they implement

Configs could diverge, but it's not an issue until you don't use something explicitly. So, the idea is to keep this `Config` as minimal and generic as possible. 

Because all of our runtimes are expected to be the "same" application, most architectural changes should be the same (pallets depending on the same pallets and things like that), which are the most important, and only minor config properties differ, which call be accessed using T::Property in most cases.

These tests also help to determine structural incoherences among our runtimes, so I think it's another good point of having them. This is how I found the line you commented on. Only properties that use other pallets are the ones that, sooner or later, should be mostly aligned.

In the worst case, where you need explicitly something different for a runtime, you can add a bound in your test:

```rust
test_foo<T: Runtime>() where T: ... {

}
```

But if you need this, probably you can test directly with the original runtime because such runtime behaves differently, and you expect something different.